### DISCUSSION TITLE
PorousFlow issues when running in parallel

### DISCUSSION THREAD

**Discussion by lloyd-5731** [idaholab/moose#24121] [Q&A Modules: General] [answered]
Hi there,

I am experiencing issues when running in parallel on our cluster.
Some of the simple models in this porous flow example are exhibiting strangely shaped flow fronts. The ideal case is two approaching vertical flow fronts.
I believe that this is linked to issue #22478 that @GiudGiud  and I were looking a

I have completely stripped back the input file to a simple Porous Flow MWE, and it is still exhibiting the same issues.

A few more things:
- The strange behaviour is only happening in parallel
- Replicated and distributed give different results
- The strange behaviour is not obvious on simpler meshes (hence this is an MWE)
- Also that I'm using a custom capillary pressure formulation which enables a distinct flow front

Has anyone got any ideas what might be causing this
<img width="884" alt="Screenshot 2023-04-19 at 15 40 17" src="https://user-images.githubusercontent.com/116571080/233110875-efa9b815-601e-4da1-8c4b-54b4506e99a3.png">
t.
^this example has fluid being pushed in from both the left and the right boundary. The flow from the right side seems normal, from the left the flow front seems incorrect however. This example is run on 48 cores.
<details>
<summary>
Here is my input file:
</summary>

```

# 2D single phase unsaturated porous flow 
# All units in SI unless otherwise stated

P0 = 101e3      # Injection pressure
rho = 920       # Density
k = 2.5e-11       # Permeability
phi = 0.45      # Porosity
T = 403
mu = 1

[Mesh]
    [plate]
        type = GeneratedMeshGenerator
        dim = 3
        xmax = 0.813
        ymax = 0.559
        zmax = 0.0024
        nx = 40
        ny = 30
        nz = 1
        elem_type = TET4
    []
[]


[UserObjects]
    [dictator]  
        type = PorousFlowDictator
        porous_flow_vars = porepressure
        number_fluid_phases = 1
        number_fluid_components = 1
    []
    [pc]
        type = CustomCapillaryPressure
        p_max = 1000
        epsilon = 80
        log_extension = false
    []
[]

[GlobalParams]
    PorousFlowDictator = dictator
    gravity = '0 0 0'
[]

[Variables]
    [porepressure]
        initial_condition = 0
    []

[]

[Kernels]
    [dmdt]
        type = PorousFlowMassTimeDerivative
        variable = porepressure
    []
    [flux]
        type = PorousFlowAdvectiveFlux
        variable = porepressure
    []
[]

[AuxVariables]
    [saturation]
        family = MONOMIAL
        order = CONSTANT
    []
[]

[AuxKernels]
    [saturation]
        type = PorousFlowPropertyAux
        variable = saturation
        property = saturation
    []
[]

[BCs]
    [injection]
        type = DirichletBC
        variable = porepressure
        boundary = 'left right'
        value = ${P0}
    []
[]

[Modules]
    [FluidProperties]
        [liquid]
            type = SimpleFluidProperties
            density0 = ${rho}
            viscosity = ${mu}
        []
    []
[]

[Materials]

    [porosity]
        type = PorousFlowPorosity
        porosity_zero = ${phi}
    []
    [permeability]
        type = PorousFlowPermeabilityConst
        permeability = '${k} 0 0   0 ${k} 0   0 0 ${k}'
    []
    [saturation_calculator]
        type = PorousFlow1PhaseP
        porepressure = porepressure
        capillary_pressure = pc
    []
    [temperature]
        type = PorousFlowTemperature
        temperature = ${T}
    []
    [massfrac]
        type = PorousFlowMassFraction
    []
    [liquid]
        type = PorousFlowSingleComponentFluid # LSPorousFlowSingleComponentFluid
        fp = liquid
        phase = 0
    []
    [relperm]
        type = PorousFlowRelativePermeabilityCorey
        n = 3
        s_res = 0.1
        sum_s_res = 0.1
        phase = 0
    []
[]

[Preconditioning]
    [smp]
        type = SMP
        full = true
        
        petsc_options_iname = '-pc_type -pc_factor_mat_solver_package'
        petsc_options_value = ' lu       mumps'
    []
    
[]

[Executioner]
    type = Transient
    solve_type = 'NEWTON'
    end_time = 20000

    [TimeStepper]
        type = SolutionTimeAdaptiveDT
        dt = 1
    []
    automatic_scaling = true
    nl_abs_tol = 2e-12 
[]

[Outputs]
    exodus = true
    print_linear_residuals = false
    file_base = "MyModel-${jobnumber}"
[]

```

</details>

<details>
<summary>
Here is my custom capillary pressure:
</summary>

```

#pragma once

#include "PorousFlowCapillaryPressure.h"
#include <cmath>

class CustomCapillaryPressure : public PorousFlowCapillaryPressure
{
public:
  static InputParameters validParams();

  CustomCapillaryPressure(const InputParameters & parameters);

  virtual Real capillaryPressureCurve(Real saturation, unsigned qp = 0) const override;
  virtual Real dCapillaryPressureCurve(Real saturation, unsigned qp = 0) const override;
  virtual Real d2CapillaryPressureCurve(Real saturation, unsigned qp = 0) const override;

  virtual Real effectiveSaturation(Real pc, unsigned qp = 0) const override;
  virtual Real dEffectiveSaturation(Real pc, unsigned qp = 0) const override;
  virtual Real d2EffectiveSaturation(Real pc, unsigned qp = 0) const override;

protected:
  const Real _p_min;
  const Real _p_max;
  const Real _epsilon;
};

```

```

#include "CustomCapillaryPressure.h"

registerMooseObject("MyApp", CustomCapillaryPressure);

InputParameters
CustomCapillaryPressure::validParams()
{
  InputParameters params = PorousFlowCapillaryPressure::validParams();
  params.addRequiredParam<Real>("p_max", "Maximum pressure at flow front (above which S=1)");
  params.addParam<Real>("p_min", 0, "Minimum pressure at flow front (below which S=0)");
  params.addParam<Real>("epsilon",
                        "Epsilon parameter, determines stretch of pressure saturation curve");
  params.addClassDescription("Custom pressure saturation relationship");
  return params;
}

CustomCapillaryPressure::CustomCapillaryPressure(const InputParameters & parameters)
  : PorousFlowCapillaryPressure(parameters),
    _p_min(getParam<Real>("p_min")),
    _p_max(getParam<Real>("p_max")),
    _epsilon(getParam<Real>("epsilon"))
{
}

Real
CustomCapillaryPressure::capillaryPressureCurve(Real saturation, unsigned /*qp*/) const
{
  return 0;
}

Real
CustomCapillaryPressure::dCapillaryPressureCurve(Real saturation, unsigned /*qp*/) const
{
  return 0;
}

Real
CustomCapillaryPressure::d2CapillaryPressureCurve(Real saturation, unsigned /*qp*/) const
{
  return 0;
}

Real
CustomCapillaryPressure::effectiveSaturation(Real pc, unsigned /*qp*/) const
{

  const float p0 = _p_min + (_p_max - _p_min) / 2;
  return 1 / (1 + std::exp((p0 - pc) / _epsilon));
}

Real
CustomCapillaryPressure::dEffectiveSaturation(Real pc, unsigned /*qp*/) const
{
  const float p0 = _p_min + (_p_max - _p_min) / 2;
  return (std::exp((p0 - pc) / _epsilon)) /
         (_epsilon * std::pow(std::exp((p0 - pc) / _epsilon) + 1, 2));
}

Real
CustomCapillaryPressure::d2EffectiveSaturation(Real pc, unsigned /*qp*/) const
{
  const float p0 = _p_min + (_p_max - _p_min) / 2;
  return (std::exp((2 * p0 + pc) / _epsilon) - std::exp((p0 + 2 * pc) / _epsilon)) /
         (std::pow(_epsilon, 2) * std::pow(std::exp(p0 / _epsilon) + std::exp(pc / _epsilon), 3));
}

```
</details>

**Comment by lloyd-5731** [comment]
Any Ideas?

**Reply by GiudGiud** [reply]
hello. I ll try to have a look today

**Reply by GiudGiud** [reply]
Is this what I should see in serial?

![serial](https://user-images.githubusercontent.com/25037151/234634244-a40350be-2d42-417a-9119-571853ec1d59.png)

**Reply by lloyd-5731** [reply]
<img width="891" alt="Screenshot 2023-04-27 at 14 44 11" src="https://user-images.githubusercontent.com/116571080/234881104-52ae44a9-37ea-489a-a801-fa67c80a551d.png">

Thank you @GiudGiud.
This is what I see (and expect to see) with the porepressure in serial.


<img width="908" alt="Screenshot 2023-04-27 at 14 46 06" src="https://user-images.githubusercontent.com/116571080/234881652-a2e1e1b0-f932-47da-ab3e-7bef7e654078.png">

And this is with the saturation.

**Reply by GiudGiud** [reply]
Umm I dont see that, and I dont think I modified the inputs.
Are you using an old moose or petsc by any chance?
any modifications from the inputs you posted?

**Reply by lloyd-5731** [reply]
I really appreciate you getting involved, thank you.
- Do you add the `CustomCapillaryPressure` files added?
- The Moose version is from this month.
- There are no other modifications apart from the capillary pressure.

**Reply by GiudGiud** [reply]
I did add those two files in my porous flow folder. 
Let me try again from scratch

**Reply by GiudGiud** [reply]
oh that was an artefact in paraview when looking at all the sidesets as well

**Reply by lloyd-5731** [reply]
Ok, do yo get the same results that I do for this case then?

**Reply by GiudGiud** [reply]
threads, serial and replicated meshes seem to agree.
 I m running distributed right now. What problem should I be seeing?

**Reply by GiudGiud** [reply]
looks the same so far for distributed. 
How long do I need to wait to see this image you pasted?
Run to completion at 20k seconds?

**Reply by lloyd-5731** [reply]
So when I run in parallel the flow fronts become non-straight. 
We would expect straight flow fronts like we have in serial.
The deformation of the flow front doesn't always happens and the area that deforms seems to depend on the number of cores I run on.

**Reply by lloyd-5731** [reply]
The image I posted occurs at around 14k seconds.

Could you please post your saturation output?
What kind of machine and how many cores are you running?

**Reply by GiudGiud** [reply]
ok I have not run it long enough. I ll wait a bit. I m running it on my macbook with 8 cores. 

Have you tried:
- using a higher order time integration scheme (bdf2?)
- decreasing the non linear absolute tolerance

This #core-dependent-behavior is making me think of numerical errors, especially as very many timesteps are needed so numerical errors accumulate.

**Reply by lloyd-5731** [reply]
- I not have tried the higher order integration scheme. I can give that a go.
- I have tried decreasing the non linear absolute tolerance, and it doesn't seem to make the issues go away. Currently the nl_abs_tol = 2e-12 which is already small by my understanding?

**Reply by GiudGiud** [reply]
I have a slight asymmetry at 3k seconds
<img width="843" alt="Screen Shot 2023-04-27 at 9 15 40 AM" src="https://user-images.githubusercontent.com/25037151/234907819-e48c83c0-8054-46ff-bf50-4c25aa90e860.png">

**Reply by lloyd-5731** [reply]
That is exactly the kind of thing I have been getting (with varying magnitudes of the asymmetry).

**Reply by lloyd-5731** [reply]
What would you recommend going forward?

**Reply by GiudGiud** [reply]
Results seem symmetric for higher order time integration

it dies when the medium is fully saturated but it might just be because it s pretty converged and it cant converge it more

<img width="843" alt="Screen Shot 2023-04-27 at 10 04 34 AM" src="https://user-images.githubusercontent.com/25037151/234920775-a018a735-a98d-4c95-93cb-47887d03a7e6.png">

**Reply by lloyd-5731** [reply]
Ok great. Thanks for the help Guillaume!

- Please can you paste the changes to the input file you have made so I can replicate these results?

Have a great day, 
Lloyd

**Reply by GiudGiud** [reply]
in executioner:
`  scheme = 'bdf2'`

run with 8 cores and distributed - mesh.
Though hopefully this no longer matters

**Reply by WilkAndy** [reply]
Thank you for getting to the bottom of this @GiudGiud .   I haven't been contributing because i thought it was probably @lloyd-5731 's computer setup that was somehow wrong, which is beyond me.  But, wow, it seems as if the accumulation of precision loss and/or numerical errors have caused this, which is something that future MOOSE users could find illuminating, thank you.

@lloyd-5731 - i wonder whether your model is somehow a bit unusual.  Many simple PorousFlow models are resistant to tiny discrepancies in porepressure (eg, diffusion just smooths all the inaccuracies) but yours isn't.  It's like there is an inherent instability in your problem - perhaps mathematically the symmetric solution is actually unstable, or something similar.

**Reply by lloyd-5731** [reply]
I have given it another go with the `scheme = 'bdf2'`, and I'm still suffering from some flow front issues here @GiudGiud.
<img width="584" alt="Screenshot 2023-04-28 at 10 18 30" src="https://user-images.githubusercontent.com/116571080/235108385-4f507d24-1aaa-4722-9ca5-72711568b668.png">
This was run on an unreasonable number of cores in this case (48) to just see what the limit is.

When I run it on 8 cores however much more normal behaviour is observed. 
 
<img width="756" alt="Screenshot 2023-04-28 at 10 30 47" src="https://user-images.githubusercontent.com/116571080/235111288-b0c004a1-6fdf-476e-badf-42d6d7e06488.png">

Could this second issue be to do with trying to use too much compute? Is that possible?

**Reply by GiudGiud** [reply]
I dont think "compute" is an issue.

Now that we know that it s likely a numerical stability issue, we need to focus on making sure the problem is tightly converged. 
I would recommend using a direct solve (LU preconditioning) and seeing if results improve.
One big difference between serial, 8 and 48 cores is the preconditioning. The default preconditioners and what they do even for a given one varies based on the number of processes

**Reply by lloyd-5731** [reply]
Ok that sounds really positive. 
I think I am already using `lu`, with my preconditioning block:

```
[Preconditioning]
    [smp]
        type = SMP
        full = true
        
        petsc_options_iname = '-pc_type -pc_factor_mat_solver_package'
        petsc_options_value = ' lu       mumps'
    [] 
[]
```

Does this look right?

**Reply by GiudGiud** [reply]
ah yes. I had seen that already. 
Umm I m not sure then. 

Do you get good performance out of openmp threads? That could be another way around the issue, doing 8 mpi * 6 threads = 48 cores

**Reply by lloyd-5731** [reply]
I will check the performance over the weekend and get back to you.
Is there a way of testing/quantifying how tightly a solve is converged?

**Reply by GiudGiud** [reply, marked as answer]
It s tough because you are also using automatic_scaling. So the convergence is achieved only up to the scaling factor basically, especially since you only have a single variable
Maybe if `compute_scaling_once = false` you ll get a more dynamic convergence criterion, but it might get too tight
